### ISSUE TITLE
AlexNet-like network loss not reducing while plain tensorflow implementation converges

### ISSUE THREAD

**Issue by chintak** [fchollet/keras#2741]
I implemented an alexnet-like network using Keras to train on 50 ImageNet classes.
Model definition:

``` python
model = Sequential()
model.add(Convolution2D(96, 11, 11, subsample=(4, 4), activation='relu',
                        border_mode='valid', name='conv_1', input_shape=(3, 256, 256)))
model.add(MaxPooling2D((3, 3), strides=(2, 2)))
model.add(ZeroPadding2D(padding=(2, 2)))
model.add(Convolution2D(256, 5, 5, activation='relu',
                        border_mode='valid', name='conv_2'))
model.add(MaxPooling2D((3, 3), strides=(2, 2)))
model.add(ZeroPadding2D(padding=(1, 1)))
model.add(Convolution2D(384, 3, 3, activation='relu',
                        border_mode='valid', name='conv_3'))
model.add(ZeroPadding2D(padding=(1, 1)))
model.add(Convolution2D(384, 3, 3, activation='relu',
                        border_mode='valid', name='conv_4'))
model.add(ZeroPadding2D(padding=(1, 1)))
model.add(Convolution2D(256, 3, 3, activation='relu',
                        border_mode='valid', name='conv_5'))
model.add(MaxPooling2D((3, 3), strides=(2, 2)))
model.add(Flatten(name='flatten'))
model.add(Dense(4096, activation='relu', name='dense_1'))
model.add(Dropout(0.5))
model.add(Dense(4096, activation='relu', name='dense_2'))
model.add(Dropout(0.5))
model.add(Dense(50, activation='softmax', name='output'))
model.compile(optimizer='sgd', loss='categorical_crossentropy',
              metrics=['accuracy'])
```

I am training it using `model.train_on_batch` method. Reading batches of data from LMDB dataset and feeding it into `train_on_batch` method.  Batch size is 64.
Observations:
- The loss is of the order ~15-16
- Seeing the network outputs after each batch using `predict_on_batch` (on training batches), shows that initially the predicted classes are random but after 2-3 batches **all predictions are of the same class** with a **probability of 1.0**.
- There is no change in the predicted class after this.
- The predicted class is random and varies from 1 trial to another.

I also implemented the **network using plain tensorflow**; model definition here - [tf.txt](https://github.com/fchollet/keras/files/268183/tf.txt). 
Observations:
- The loss starts with ~2-3.
- Trains as expected.

Any idea what am I doing incorrectly in Keras? Any tests I could run to ensure Keras is installed as expected? I'm unable to figure out why there is such a huge difference in the starting loss values. Is there a possible vanishing or exploding gradients problem here? 

System details:
- Keras v1.0.3
- Tensorflow v0.8.0rc0
- GTX 960 4GB

**Comment by joelthchao** [comment]
Some personal opinions:
1. Weight initialization migh be different. ([docs](http://keras.io/initializations/), although bias is initialized with zeros in Keras)
2. Is optimizer's learning rate the same?
3. Why don't you use `border_mode='same'` to remove `ZeroPadding2D`?

**Comment by chintak** [comment]
@joelthchao:
1. I tried with 'normal', 'uniform', 'glorot_normal' as well.
2. In plain tensorflow implementation I am using `tf.train.GradientDescentOptimizer(1e-2)` optimizer. For keras, I tried varying the learning rate from 1e-1 to 1e-5. With a smaller lr, it requires a greater number of batches to yield all predictions as the same class.
3. Yep, I had used that previously - obviously identical results.

**Comment by chintak** [comment]
I also tried switching to Theano backend to no additional success. But clearly the backend does not seem to be an issue.

**Comment by joelthchao** [comment]
Sorry, I can't see any other reasons for this problem. Do you have a runnable script with data to reproduce it? I might have time to take it a look.

**Comment by chintak** [comment]
@joelthchao, sent the script over mail.

I tried changing the activation of the 2nd dense layer to 'sigmoid' from 'relu' and surprisingly now the predicted class is changing from 1 mini-batch to another but all are changing simultaneously, i.e. still a single class predicted for the entire mini-batch.

**Comment by chintak** [comment]
I tried running the tests - `py.test tests/` and got a whole lot of `[gw24] node down: Not properly terminated` - entire [test log](http://pastebin.com/J73crdZj). Is this normal? What am I missing?

**Comment by chintak** [comment]
The problem was related to random weight initialization.

In the tensorflow implementation, I played around with different manual initializations and noticed the same "large loss and no effective training" problem when I initialized the weights with a `stddev` of 0.1. The activations in the last layer, prior to applying softmax were in the range of [-3000, 5000], resulting in nearly one-hot vectors after softmax. On reducing the `stddev` for initialization, I was able to train the network. Manual initialization in Keras, allowed me to train the model with similar performance in Keras as well.

My initializations: 
- Convolution layers: `weights ~ N(0.0, 0.01)` and `biases ~ 0.1`
- Dense layers: `weights ~ N(0.0, 0.005)` and `biases ~ 0.1`

I wonder why any of the existing initialization (mainly "glorot_uniform" and "glorot_normal") weren't able to capture this.
